sdk/src/Core/Amazon.Runtime/Internal/Util/StreamingUtf8JsonReader.cs

```diff
@@ -0,0 +1,158 @@
+                throw new ArgumentException("Stream must not be null. Please initialize a stream and pass it into the constructor.");
+
+            _stream = stream;
+            _buffer = ArrayPool<byte>.Shared.Rent(AWSConfigs.StreamingUtf8JsonReaderBufferSize ?? 4096);
+            // need to initialize the reader even if the buffer is empty because auto-default of unassigned fields is only 
+            // supported in C# 11+
+
+            int utf8BomLength = JsonConstants.Utf8Bom.Length;
+            Debug.Assert(_buffer.Length >= utf8BomLength);
+
+            int bytesRead = _stream.Read(_buffer, 0, _buffer.Length);
+            int start = 0;
+            if (_buffer.AsSpan().StartsWith(JsonConstants.Utf8Bom))
+            {
+                start += utf8BomLength;
+                bytesRead -= utf8BomLength;
+            }
+
+            _reader = new Utf8JsonReader(_buffer.AsSpan(start, bytesRead), isFinalBlock: bytesRead == 0, new JsonReaderState(_jsonOptions));
+        }
+
+        /// <summary>
+        /// Reads data from the buffer and fetches more data from the stream to fill the buffer it more data exists.
+        /// </summary>
+        /// <returns>true if there is more data, false otherwise.</returns>
+        public bool Read()
+        {
+            // hasMoreData can return false if the value starts in one buffer and leaks into the next buffer
+            bool hasMoreData = _reader.Read();
+
+            while (!hasMoreData)
+            {
+                GetMoreBytesFromStream(_stream, ref _buffer, ref _reader);
+
+                if (_reader.IsFinalBlock)
+                    break;
+
+                hasMoreData = _reader.Read();
+            }
+
+            return hasMoreData;
+        }
+
+        private static void GetMoreBytesFromStream(Stream stream, ref byte[] buffer, ref Utf8JsonReader reader)
+        {
+            int bytesRead = 0;
+            int previousBufferLength = buffer.Length;
+            if (reader.BytesConsumed < buffer.Length)
+            {
+                ReadOnlySpan<byte> leftover = buffer.AsSpan().Slice((int)reader.BytesConsumed);
+                // If BytesConsumed is 0 that means that the previous Read failed because the JSON token was too large to fit in the buffer.
+                // In that case we need to resize the buffer and try again to read the JSON token.
+                if (reader.BytesConsumed == 0)
+                {
+                    var resizedBuffer = ArrayPool<byte>.Shared.Rent(Math.Min(int.MaxValue, (buffer.Length * 2)));
+                    Logger.GetLogger(typeof(StreamingUtf8JsonReader)).DebugFormat("Resizing buffer from {0} to {1}", buffer.Length, resizedBuffer.Length);
+
+                    buffer.AsSpan().CopyTo(resizedBuffer);
+                    ArrayPool<byte>.Shared.Return(buffer);
+                    buffer = resizedBuffer;
+                    bytesRead = FillBuffer(stream, ref buffer, leftover.Length, buffer.Length - leftover.Length);
```

Comment:
Since this is a brand new empty buffer why would would the offset ever be anything then `0` and bytes to read be `buffer.Length`? Then you can move the `leftover` variable to in the else block the only place it is used.

Reply:
It isn't a brand new empty buffer since we do 
```
buffer = resizedBuffer;
```
The buffer is basically resizedBuffer with its first half filled with buffer's data. So, the offset needs to be `leftover.Length` in this case because we don't want to overwrite that data, which is what would happen if we set the offset to 0.

leftover's length is also important because when we do
```
buffer = resizedBuffer
```

we lose the length of the buffer (since it is now buffer.Length * 2). So that is how we keep track of the offset as well as how much data to read. And since we use leftover.Length in the other else block as well, I'd rather just keep it there. It doesn't allocate a new buffer, since we're calling `AsSpan()`, and 

I guess since reader.BytesConsumed = 0. I can get rid of perviousBufferLength though, since leftover.Length == previousBufferLength at this point.

I will add more comments, as it is nuanced